core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala

```diff
@@ -58,6 +74,11 @@ class ZkMetadataCache(brokerId: Int) extends MetadataCache with Logging {
   this.logIdent = s"[MetadataCache brokerId=$brokerId] "
   private val stateChangeLogger = new StateChangeLogger(brokerId, inControllerContext = false, None)
 
+  // Features are updated via ZK notification (see FinalizedFeatureChangeListener)
+  @volatile private var featuresAndEpoch: Option[FinalizedFeaturesAndEpoch] = Option.empty
+  private val featureLock = new ReentrantLock()
+  private val featureCond = featureLock.newCondition()
```

Comment:
In FinalizedFeatureCache the same volatile shown here was used, but it was using `synchronized` blocks and wait/notify. I opted for an explicit lock and condition to avoid locking on the whole class